packages/web-api/src/services/preApproval.ts

```diff
@@ -0,0 +1,379 @@
+        email: input.email.toLowerCase().trim(),
+        roles: input.roles,
+        created_by_user_id: input.createdByUserId
+      },
+      include: {
+        created_by_user: {
+          select: {
+            id: true,
+            email: true
+          }
+        }
+      }
+    });
+  }
+
+  /**
+   * Retrieves multiple pre-approved users with optional filtering
+   *
+   * @param options - Query options for filtering and pagination
+   * @returns Promise<PreApprovedUser[]> - Array of matching pre-approved users
+   */
+  async getMany(options: GetPreApprovedUsersOptions = {}): Promise<PreApprovedUser[]> {
+    const { email, used, createdByUserId, limit = 50, offset = 0 } = options;
+
+    return await this.prisma.preApprovedUser.findMany({
+      where: {
+        ...(email && { email: { contains: email.toLowerCase(), mode: 'insensitive' } }),
+        ...(used !== undefined && { used }),
+        ...(createdByUserId && { created_by_user_id: createdByUserId })
+      },
+      include: {
+        created_by_user: {
+          select: {
+            id: true,
+            email: true
+          }
+        }
+      },
+      orderBy: {
+        created_at: 'desc'
+      },
+      take: Math.min(limit, 100), // Cap at 100 for performance
+      skip: offset
+    });
+  }
+
+  /**
+   * Gets count of pre-approved users matching the given criteria
+   *
+   * @param options - Query options for filtering
+   * @returns Promise<number> - Count of matching records
+   */
+  async getCount(options: GetPreApprovedUsersOptions = {}): Promise<number> {
+    const { email, used, createdByUserId } = options;
+
+    return await this.prisma.preApprovedUser.count({
+      where: {
+        ...(email && { email: { contains: email.toLowerCase(), mode: 'insensitive' } }),
+        ...(used !== undefined && { used }),
+        ...(createdByUserId && { created_by_user_id: createdByUserId })
+      }
+    });
+  }
+
+  /**
+   * Updates an existing pre-approved user
+   *
+   * @param id - The ID of the pre-approved user to update
+   * @param input - The updated data
+   * @returns Promise<PreApprovedUser> - The updated pre-approved user
+   * @throws Error if pre-approval not found or already used
+   */
+  async update(id: number, input: UpdatePreApprovedUserInput): Promise<PreApprovedUser> {
+    // Check if pre-approval exists and is not used
+    const existing = await this.prisma.preApprovedUser.findUnique({
+      where: { id }
+    });
+
+    if (!existing) {
+      throw new NotFoundException(`Pre-approved user with ID ${id} not found`);
+    }
+
+    if (existing.used) {
+      throw new BadRequestException('Cannot update a pre-approval that has already been used');
+    }
+
+    // If updating email, check for conflicts
+    if (input.email && input.email.toLowerCase().trim() !== existing.email) {
+      const emailConflict = await this.prisma.preApprovedUser.findUnique({
+        where: { email: input.email.toLowerCase().trim() }
```

Comment:
you're repeating `.toLowerCase().trim()` 3 times in this function, better to create variable. Also, arguably should create a function to standardize this though it's just 2 calls.